daml-lf/transaction/src/main/scala/com/daml/lf/TransactionNodesStatistics.scala

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf
+package transaction
+
+/** Container for transaction statistics.
+  *
+  * @param creates number of creates nodes,
+  * @param consumingExercisesByCid number of consuming exercises by contract ID nodes,
+  * @param nonconsumingExercisesByCid number of non-consuming Exercises by contract ID nodes,
+  * @param consumingExercisesByKey number of consuming exercise by contract key nodes,
+  * @param nonconsumingExercisesByKey number of non-consuming exercise by key nodes,
+  * @param fetchesByCid number of fetch by contract ID nodes,
+  * @param fetchesByKey number of fetch by key nodes,
+  * @param lookupsByKey number of lookup by key nodes,
+  * @param rollbacks number of rollback nodes.
+  */
+final case class TransactionNodesStatistics(
+    creates: Int,
+    consumingExercisesByCid: Int,
+    nonconsumingExercisesByCid: Int,
+    consumingExercisesByKey: Int,
+    nonconsumingExercisesByKey: Int,
+    fetchesByCid: Int,
+    fetchesByKey: Int,
+    lookupsByKey: Int,
+    rollbacks: Int,
+) {
+
+  def +(that: TransactionNodesStatistics) =
+    TransactionNodesStatistics(
+      creates = this.creates + that.creates,
+      consumingExercisesByCid = this.consumingExercisesByCid + that.consumingExercisesByCid,
+      nonconsumingExercisesByCid =
+        this.nonconsumingExercisesByCid + that.nonconsumingExercisesByCid,
+      consumingExercisesByKey = this.consumingExercisesByKey + that.consumingExercisesByKey,
+      nonconsumingExercisesByKey =
+        this.nonconsumingExercisesByKey + that.nonconsumingExercisesByKey,
+      fetchesByCid = this.fetchesByCid + that.fetchesByCid,
+      fetchesByKey = this.fetchesByKey + that.fetchesByKey,
+      lookupsByKey = this.lookupsByKey + that.lookupsByKey,
+      rollbacks = this.rollbacks + that.rollbacks,
+    )
+
+  def exercisesByCid: Int = consumingExercisesByCid + nonconsumingExercisesByCid
+  def exercisesByKey: Int = consumingExercisesByKey + nonconsumingExercisesByKey
+  def exercises: Int = exercisesByCid + exercisesByKey
+  def consumingExercises: Int = consumingExercisesByCid + consumingExercisesByKey
+  def nonconsumingExercises: Int = nonconsumingExercisesByCid + nonconsumingExercisesByKey
+  def fetches: Int = fetchesByCid + fetchesByKey
+  def byKeys: Int = exercisesByKey + fetchesByKey + lookupsByKey
+  def actions: Int = creates + exercises + fetches + lookupsByKey
+  def nodes: Int = actions + rollbacks
+}
+
+object TransactionNodesStatistics {
+
+  val Empty = TransactionNodesStatistics(0, 0, 0, 0, 0, 0, 0, 0, 0)
+
+  private[this] val numberOfFields = Empty.productArity
+
+  private[this] val Seq(
```

Comment:
@cocreature wrote:
> I’m confused, a single traversal is single threaded. Each traversal acquires new state. Which part are you worried about?

I am very confuse by your comment. I do not understand where your concern about concurrency come from. 
The `xxxIdx` values will be instanciate once and only once for each JVM run. 

Here I just want to avoid to give the explicit "Index" of each field of the case class `TransactionNodesStatistics`. 
It will also crash if somebody add a field the the case class but do not add an "index".